[reland 2] Call jit decomp in VariableType to improve forward AD coverage

soulitzer opened the pull request:
Reland of https://github.com/pytorch/pytorch/pull/84675

soulitzer committed:
Revert "Revert "Call jit decomposition in VariableType to increase forward AD coverage (#84151)""

This reverts commit acb4a09628284201281e262aaee58e3dc6be9c2b.

ghstack-source-id: 387eda832cbd9d976acbbbb9cf5468f8e0a9b0a9
Pull Request resolved: https://github.com/pytorch/pytorch/pull/84675

soulitzer committed:
Access JIT decomposition utils through an interface

soulitzer commented:
Hopefully internal builds should passing now. The issue was that certain builds (e.g. lite_trainer) depended on the VariableType files, but do not depend on the jit. Including all of those additional dependencies to get jit to work on lite trainer will blow up the binary size so I don't think we want that.

Instead, what I'm doing in this updated PR (see latest commit for new changes) is the same thing that is done with VariableHooksInterface, i.e. provide the apis conditionally depending on whether the library depends on jit.

soulitzer commented on torch/csrc/autograd/VariableTypeUtilsDependOnOps.h:
```diff
@@ -0,0 +1,60 @@
+// native_functions.yaml meaning the file will need to be re-compiled every time
+// an operator is changed or added. We cannot simply put these functions in
+// VariableType.h and VariableTypeutils.h, since they are included in files like
+// ADInplaceOrViewType_X.cpp which don't always want to be recompiled.
```
This message might be out of date now, we could probably merge this file back into VariableTypeUtils.h

soulitzer replied:
This file used to depend on decomposition_registry directly which depends on more jit things which depend on interned strings.

torch/csrc/jit/ir/ir.h -> aten/src/ATen/core/interned_strings.h

soulitzer commented on torch/csrc/jit_decomp_interface.cpp:
```diff
@@ -0,0 +1,23 @@
+JitDecompInterface* getJitDecomp() {
+  TORCH_CHECK(
+      fns,
+      "Support for JIT decompositions has not been loaded; have you linked against TBD?")
```
This message should be more informative. I'll probably update this and other things once I'm sure internal builds pass.

soulitzer commented on torch/csrc/jit_decomp_interface.h:
```diff
@@ -0,0 +1,31 @@
+#include <c10/macros/Export.h>
+
+namespace torch {
+namespace autograd {
```
Since this is in the autograd namespace, maybe I want to move this file in to the torch/csrc/autograd directory

soulitzer replied:
No longer doing this

zou3519 commented on torch/csrc/jit/runtime/decomposition_registry.cpp:
```diff
@@ -160,6 +161,46 @@ void RegisterDecomposition(
 }
 
+struct JitDecomp final : torch::autograd::impl::JitDecompInterface {
+  bool has_jit_decomposition_(const c10::FunctionSchema& schema) const override;
+  void run_jit_decomposition_(
+      const c10::OperatorHandle& op,
+      torch::jit::Stack* stack) const override;
```
Why do these functions end in '_'?

soulitzer replied:
If the member function were also named `run_jit_decomposition` that would hide names outside of the class scope.

Alternatively, I can specify the function with its namespace or use a using declaration.

soulitzer replied:
Updated to specify the namespace to get around name hiding

zou3519 commented on torch/csrc/jit_decomp_interface.h:
```diff
@@ -0,0 +1,31 @@
+namespace torch {
+namespace autograd {
+namespace impl {
+
+struct TORCH_API JitDecompInterface {
```
Should add some comments or docs about why this is necessary and what it is used for

zou3519 reviewed (commented):
The code and approach looks good to me. Two main comments:
- There's a question of what we want to do in the long term. Should we take the TorchScript dependency, or should we just invoke the decomposition in Python? Since the TorchScript route is easy to do and proven to work, we should go with this PR for the short-to-medium term.
- It took me a while to figure out what JitDecompInterface was and why we needed it. We should put a comment with a NOTE somewhere so that folks coming along to read it aren't confused.

soulitzer committed:
Clean up and address comments

soulitzer commented:
I've confirmed that internal tests are passing now. I've cleaned up this PR a little bit and addressed the comments, so it should be ready for another look!

> There's a question of what we want to do in the long term. Should we take the TorchScript dependency, or should we just invoke the decomposition in Python? Since the TorchScript route is easy to do and proven to work, we should go with this PR for the short-to-medium term.

This route makes sense to me as well

zou3519 commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -442,5 +446,56 @@ inline std::vector<c10::ScalarType> to_args_scalartypes(
+
+namespace {
+
+// if run_jit_decomposition were not a member function, we wouldn't have to
+// use a functor
```
Can you elaborate a bit more?

zou3519 commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -442,5 +446,56 @@ inline std::vector<c10::ScalarType> to_args_scalartypes(
+
+// if run_jit_decomposition were not a member function, we wouldn't have to
+// use a functor
+class MyFunctor final : public c10::OperatorKernel {
```
Please rename MyFunctor to something more semantically meaningful

zou3519 commented on torch/csrc/jit_decomp_interface.cpp:
```diff
@@ -0,0 +1,21 @@
+#include <torch/csrc/jit_decomp_interface.h>
+
+namespace torch {
+namespace autograd {
+namespace impl {
+
+namespace {
+JitDecompInterface* impl = nullptr;
+}
+
+void setJitDecompImpl(JitDecompInterface* impl_) {
+  impl = impl_;
+}
+
+JitDecompInterface* getJitDecompImpl(c10::string_view name) {
+  return impl;
+}
+
```
Based on the namespace I would have expected this file to go under torch/csrc/autograd

zou3519 commented on torch/csrc/jit_decomp_interface.h:
```diff
@@ -0,0 +1,54 @@
+
+// NOTE: [Jit Decomposition Interface]
+//
+// For some context of why we need this at all, see NOTE: [forward-mode AD
+// decompositions hack]
```
Why are we calling this a hack? Since we're upstreaming the mechanism to PyTorch, I would consider it now a supported alternate method to get forward-mode AD coverage. If we want to keep calling it a hack, then we should align on what needs to happen for it to not be a hack

soulitzer replied:
This is a very good point, I don't think we have plans to replace it anytime soon.

zou3519 reviewed (approved):
LGTM, minor comments

soulitzer committed:
Address comments

soulitzer commented:
@pytorchbot merge -g
